ompi/mca/coll/adapt/coll_adapt_ireduce.c

```diff
@@ -0,0 +1,929 @@
+    ompi_request_complete(temp_req, 1);
+    return OMPI_SUCCESS;
+}
+
+/*
+ * Callback function of isend
+ */
+static int send_cb(ompi_request_t * req)
+{
+    ompi_coll_adapt_reduce_context_t *context =
+        (ompi_coll_adapt_reduce_context_t *) req->req_complete_cb_data;
+    OPAL_OUTPUT_VERBOSE((30, mca_coll_adapt_component.adapt_output,
+                         "[%d]: ireduce_send_cb, peer %d, seg_id %d\n", context->con->rank,
+                         context->peer, context->frag_id));
+    int err;
+
+    opal_atomic_sub_fetch_32(&(context->con->ongoing_send), 1);
+
+    /* Send a new segment */
+    OPAL_THREAD_LOCK(context->con->mutex_recv_list);
+    ompi_coll_adapt_item_t *item =
+        get_next_ready_item(context->con->recv_list, context->con->tree->tree_nextsize);
+    OPAL_THREAD_UNLOCK(context->con->mutex_recv_list);
+
+    if (item != NULL) {
+        /* Get new context item from free list */
+        ompi_coll_adapt_reduce_context_t *send_context =
+            (ompi_coll_adapt_reduce_context_t *) opal_free_list_wait(mca_coll_adapt_component.
+                                                                    adapt_ireduce_context_free_list);
+        if (context->con->tree->tree_nextsize > 0) {
+            send_context->buff = context->con->accumbuf[item->id];
+        } else {
+            send_context->buff =
+                context->buff + (item->id - context->frag_id) * context->con->segment_increment;
+        }
+        send_context->frag_id = item->id;
+        send_context->peer = context->peer;
+        send_context->con = context->con;
+        OBJ_RETAIN(context->con);
+
+        opal_atomic_add_fetch_32(&(context->con->ongoing_send), 1);
+
+        int send_count = send_context->con->seg_count;
+        if (item->id == (send_context->con->num_segs - 1)) {
+            send_count = send_context->con->count - item->id * send_context->con->seg_count;
+        }
+
+        OPAL_OUTPUT_VERBOSE((30, mca_coll_adapt_component.adapt_output,
+                             "[%d]: In send_cb, create isend to seg %d, peer %d, tag %d\n",
+                             send_context->con->rank, send_context->frag_id, send_context->peer,
+                             send_context->con->ireduce_tag - send_context->frag_id));
+
+        ompi_request_t *send_req;
+        err =
+            MCA_PML_CALL(isend
+                         (send_context->buff, send_count, send_context->con->datatype,
+                          send_context->peer,
+                          context->con->ireduce_tag - send_context->frag_id,
+                          MCA_PML_BASE_SEND_SYNCHRONOUS, send_context->con->comm, &send_req));
+        if (MPI_SUCCESS != err) {
+            return err;
+        }
+
+        /* Release the item */
+        OBJ_RELEASE(item);
+
+        /* Set the send callback */
+        ompi_request_set_callback(send_req, send_cb, send_context);
+    }
+
+    OPAL_THREAD_LOCK(context->con->mutex_num_sent);
```

Comment:
Is that mutex really needed? It seems like an atomic operation would do just fine here

Reply:
I think that technically you are correct, we could have done it with an atomic operation instead. We will revisit this code in the future and clean such cases, but I don't think it has any impact neither on correctness nor performance.